ksqldb-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java

```diff
@@ -196,6 +197,7 @@ static Expression rewriteTimeComparisonForFilter(final Expression expression) {
       final SchemaKStream<K> otherSchemaKStream,
       final ColumnName keyColName,
       final JoinWindows joinWindows,
+      final Optional<WindowTimeClause> gracePeriod,
```

Comment:
Why can't we just retrieve the grace period from the JoinWindows? I.e. we just set the grace time in JoinWindows if `GRACE` specified, and then pass around that object only.

Reply:
EDIT: I knew this is only used for `WITH` and the `GRACE PERIOD` is out of the scope; I guess my meta question is: why can't we have the `GRACE PERIOD` captured as optional in `WithinExpression` as well?

Reply:
The `JoinWindows` has a default grace (24h) if it is not passed one. I cannot know if the grace was optional in the ksql syntax with that default. Say a user set GRACE PERIOD 24 HOURS, then `JoinWindows` will be set to 24h, which it could be the default (no fix enabled) or manually set (fix enabled).

The grace is optional in the `WithinExpression`. I could pass the `WithinExpression` all down to the `StreamStreamJoinBuilder` but I would need to add the `ksqldb-parser` dependency into the `ksqldb-streams` module.

Reply:
Let's say in the future if we do not need to rely on GRACE to enable the fix or not, then would this code be changed again to only pass in the `JoinWindows` object still?

If yes, I'd suggest we just keep doing that right away: we could, e.g., a wild idea is to have an extended JoinWindows that adds a flag indicating if user specified grace, and then pass around that object from WithinExpression -> ExecutionStepFactory -> StreamStreamJoin, and let it to return a `graceSpecified` rather than `join.getGraceMillis()`, used by the `StreamStreamJoinBuilder`. In the future we can just remove the extended class and the `graceSpecified` from StreamStreamJoin. Does that work?

Reply:
I could do that. But it would only be used to pass the window+graceSpecified from `SchemaKStream#leftJoin()` -> `ExecutionStepFactory#streamStreamJoin()`. After that, the rest should be passed as millis to the `StreamStreamJoin` because of compatibility. Look at the `StreamStreamJoin` class. It has a constructor that is used by a JSON mapper when reads a plan. That already accepts before/after millis. I cannot change that constructor to pass the `JoinWindows`.

I'm not sure if it is worth creating a class around the JoinWindows for one single change and then remove it. I think we can remove the gracePeriod later easier than removing a class.

Reply:
Okay, I think I'm convinced. Let's just keep it as is then.